packages/puppeteer-core/src/common/bidi/Page.ts

```diff
@@ -549,6 +550,224 @@ export class Page extends PageBase {
     }
   }
 
+  #exposedFunctionId = Math.random();
+  override async exposeFunction<Args extends unknown[], Ret>(
+    name: string,
+    fn:
+      | ((...args: Args) => Awaitable<Ret>)
+      | {default: (...args: Args) => Awaitable<Ret>}
+  ): Promise<void> {
+    if ('default' in fn) {
+      fn = fn.default;
+    }
+    const func = fn;
+
+    const argsChannel = `__puppeteer__page_exposeFunction_${
+      this.#exposedFunctionId
+    }_args`;
+    const resolveChannel = `__puppeteer__page_exposeFunction_${
+      this.#exposedFunctionId
+    }_resolve`;
+    const rejectChannel = `__puppeteer__page_exposeFunction_${
+      this.#exposedFunctionId
+    }_reject`;
+    ++this.#exposedFunctionId;
+
+    type SendArgsChannel = (value: [id: number, args: Args]) => void;
+    type SendResolveChannel = (
+      value: [id: number, resolve: (ret: FlattenHandle<Awaited<Ret>>) => void]
+    ) => void;
+    type SendRejectChannel = (
+      value: [id: number, reject: (error: unknown) => void]
+    ) => void;
+
+    interface RemotePromiseCallbacks {
```

Comment:
This is a reasonable use case. We use the types here only, so it wouldn't be reasonable to abstract it to the top of the class where onlookers would question where it is used. Placing them here makes it obvious.

Reply:
I disagree as this way we bloat the function (implementation) with types that most (if not all) IDE will show you good information about. That way you can focus on the code part rather then the type part.

Reply:
let's move the implementation in a standalone class?

Reply:
`class ExposedFunction { constructor(connection, browsingContext); }`

Reply:
as exposing functions should be possible for any contexts theoretically?

Reply:
I think leaving it as is would make more sense. This code isn't used anywhere else so the gain of using a class is moot.

Reply:
the current change is lengthy and deeply nested with the use of inner types and inner callbacks so moving it to a separate class would allow binding the callbacks to the class instance, simplifying the callback subscription management and moving the types to the top-level provided that the class is placed in its own file (reducing nesting).

Reply:
Alternatively, we can also start implementing `unexposeFunction` here and create another PR that implements it on the page, however I'd rather scope this PR to just `exposeFunction`.

Reply:
sgtm to add a todo, but let's make sure tests pass before landing.